src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/csv/ReadCSVStep.java

```diff
@@ -0,0 +1,197 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.pipeline.utility.steps.csv;
+
+import org.apache.commons.csv.CSVFormat;
+import hudson.Extension;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import javax.annotation.Nonnull;
+import org.codehaus.groovy.control.customizers.ImportCustomizer;
+import org.jenkinsci.plugins.pipeline.utility.steps.AbstractFileOrTextStep;
+import org.jenkinsci.plugins.pipeline.utility.steps.AbstractFileOrTextStepDescriptorImpl;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.Whitelist;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.cps.GroovyShellDecorator;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+/**
+ *
+ * @author Stuart Rowe
+ */
+public class ReadCSVStep extends AbstractFileOrTextStep {
+
+    private static final String ORG_APACHE_COMMONS_CSV = "org.apache.commons.csv";
+    private static final String[] ORG_APAHACE_COMMONS_CSV_WHITELISTED_CLASSES = {"CSVFormat", "CSVRecord"};
+    private CSVFormat format;
+
+    @DataBoundConstructor
+    public ReadCSVStep() {
+        this.format = CSVFormat.DEFAULT;
+    }
+
+    public CSVFormat getFormat() {
+        return this.format;
+    }
+
+    @DataBoundSetter
+    public void setFormat(CSVFormat format) {
+        this.format = format;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new ReadCSVStepExecution(this, context);
+    }
+
+    @Extension
+    public static class DescriptorImpl extends AbstractFileOrTextStepDescriptorImpl {
+
+        public DescriptorImpl() {
+
+        }
+
+        @Override
+        public String getFunctionName() {
+            return "readCSV";
+        }
+
+        @Override
+        @Nonnull
+        public String getDisplayName() {
+            return Messages.ReadCSVStep_DescriptorImpl_displayName();
+        }
+    }
+    
+    /**
+     * Auto imports org.apache.commons.csv.* .
+     */
+    @Extension(optional = true)
+    public static class PackageAutoImporter extends GroovyShellDecorator {
+        @Override
+        public void customizeImports(CpsFlowExecution context, ImportCustomizer ic) {
+            ic.addStarImports(ORG_APACHE_COMMONS_CSV);
+        }
+    }
+
+    /**
+     * Whitelists all non static setters, getters and constructors in the package org.apache.commons.csv.
+     * As well as static fields.
+     */
+    @Extension
+    public static class WhiteLister extends Whitelist {
```

Comment:
I don't know enought about the `org.apache.commons.csv` library to know if it is safe to whitelist the methods and fields like this.

Reply:
I'm not particularly familiar with `org.apache.commons.csv` either.  At a minimum, methods of `CSVRecord` should be whitelisted to allow users to interact with the parsed csv data.  A `List<CSVRecord>` is much nicer to work with than a `List<Map<String,String>>`, but that is also an option if there is still concern. See: [CSVRecord](https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVRecord.html).

I'll revisit which methods of `CSVFormat` are being whitelisted here. There are convenience methods for printing that should probably be restricted. Only the static fields for predefined formats, the static creation methods and the `with*` methods are really needed. See: [CSVFormat](https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html).